### ISSUE TITLE
Severe TPU/CPU behaviour discrepency

### ISSUE THREAD

**Issue by MikeOfZen** [tensorflow/tensorflow#34932]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):No
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:No
- TensorFlow installed from (source or binary): pip
- TensorFlow version (use command below): 2.1.0-dev20191203
- Python version: 3.5
- GPU model and memory: TPU (nightly-2.x)

**Describe the current behavior**
When training using a TPU backend, if a `@tf.function` `function` code is defined before connecting to a TPU cluster. Calling the function (as `Dataset.map(function)`) results in a python kernel crash, without any errors or other info. 

This is especially severe (IMHO), since, as the training code base grows, more autograph functions are defined in modules instead of on the main program. As it's natural to import modules at the start of the program, if the TPU connection is initiated after the imports, using the @tf.fucntion code defined in the modules results in a kernel crash.
If the same code is, however, being run just on the CPU, it works as expected.
This leads to a somewhat frustrating experience of everything working on a CPU dev env, and then crashing inexplicably when connecting to a TPU.
The unintuitive solution is to run the TPU connection boilerplate before any imports.

**Describe the expected behavior**
1. connecting to a TPU shouldn't create an implicit scope, if a scope is required it should be with a `with` idiom
2. It should be better documented if all autograph function definitions should be defined after connecting to a TPU
3. If a code executed within a TPU scope depends on a code defined outside, it should fail gracefully and informatively (not crash the kernel)

**Code to reproduce the issue**

---
Fail case:
```
import tensorflow as tf

@tf.function
def test_func(a):
    return a**3

train, test = tf.keras.datasets.fashion_mnist.load_data()
images, labels = train
images = images/255
ds = tf.data.Dataset.from_tensor_slices((images))

TPU_IP = '10.0.3.2' #this requires a working nightly-2.x TPU cluster (2v-8)
tpu_address = 'grpc://' + TPU_IP + ':8470'
resolver=tf.distribute.cluster_resolver.TPUClusterResolver(tpu=tpu_address)
tf.config.experimental_connect_to_cluster(resolver)
tf.tpu.experimental.initialize_tpu_system(resolver)
strategy = tf.distribute.experimental.TPUStrategy(resolver)

dsf=ds.map(test_func)
```
---
Working case:
```
import tensorflow as tf

TPU_IP = '10.0.3.2' #this requires a working nightly-2.x TPU cluster (2v-8)
tpu_address = 'grpc://' + TPU_IP + ':8470'
resolver=tf.distribute.cluster_resolver.TPUClusterResolver(tpu=tpu_address)
tf.config.experimental_connect_to_cluster(resolver)
tf.tpu.experimental.initialize_tpu_system(resolver)
strategy = tf.distribute.experimental.TPUStrategy(resolver)

train, test = tf.keras.datasets.fashion_mnist.load_data()
images, labels = train
images = images/255
ds = tf.data.Dataset.from_tensor_slices((images))

@tf.function
def test_func(a):
    return a**3

dsf=ds.map(test_func)
```
---
The reproduction code doesn't use imports, but 
```
@tf.function
def test_func(a):
    return a**3
```
Would typically run as part of the import code, and not the main program.

**Other info / logs**
No error logs produced

**Comment by byronyi** [comment]
I’m told that this issue will be resolved in a recent release of TF TPU runtime, probably next week. 

Get in touch with your GCP solution or technical support team to confirm.

**Comment by MikeOfZen** [comment]
Thank you for the update!
As for myself, I can get by with using the aforementioned solution of initiating connection before imports.
But its good to know that a formal fix is coming soon.
Best regards.

**Comment by byronyi** [comment]
Sorry I did not take a closer look; your problem seems a different one. Btw I’m not related to Google so take my advice at your discretion.

**Comment by saberkun** [comment]
To my knowledge, strategy = tf.distribute.experimental.TPUStrategy(resolver) will help you get into the correct device scope because TPU is always the remote client.
@rxsang as expert on this.
Probably need creating a TPUStrategy at the beginning.

**Comment by rxsang** [comment]
The reason is because in TPU case is a 2VM environment (client + Cloud TPU), there are lots of things happening in the initialization time. E.g. in experimental_connect_to_cluster we will enter into the tpu worker device scope, because we expect everything to happen on the tpu worker rather the coordinator/client, otherwise the performance may be bad. Thus we actually expect user always put the TPUStrategy creation and initialization code in the beginning of the program. I will try to make the documentation better.

**Comment by MikeOfZen** [comment]
@rxsang , Thank you for the information.
I don't doubt that the interaction between the TPU VM and the client VM is very complicated. but that was my meaning of 'implicit' scope. 
In this scenario, I believe it would be better to force all commands to run within an explicit scope, with the pythonic `with` idiom, this greatly reduces the chance for strange errors that fall between the cracks of the 2 VMs.

**Comment by rxsang** [comment]
Your observation is correct. There are two scopes here: the strategy scope and the default device scope. For strategy scope, we expose it to the user and would like user to handle that. For default device scope, currently it is set in tf.config.experimental_connect_to_cluster(resolver), and you can always set tf.config.experimental_connect_to_cluster(resolver, make_master_device_default=False) to disable entering into scope automatically. But that means you may have to put your code in a "with tf.device("/job:worker:1"):" scope, and that has to be after initialization code as well. Because before the experimental_connect_to_cluster, you don't see any remote devices. That's why we handle it for users, because we think it is troublesome for users to handle the default device scope themselves.

I'm still not clear why this is a scope issue, my understanding is that you are trying to build something like dataset/tf.function before TPU initialization, the expectation is users should always do TPUStrategy creation and initialization first.

**Comment by MikeOfZen** [comment]
@rxsang 
I think some of it comes down to miscommunication.
I'm a relatively new TF user, joined after 2.0, so I may be unaware of some of the intricacies of TF, especially in distributed settings. I followed the documentation guides, and though I might have missed it, I haven't seen such an expectation ("the expectation is users should always do TPUStrategy creation and initialization first."). I only realized it after encountering these crashes.
While I personally would prefer an explicit scope definition if its required, I can certainly understand the design decision. but it should be better communicated that it should be done first, and before even the imports.

But the other part of the issue is that it crashes (the whole python kernel) ungracefully and without any indication of the reason.
If I had some exception information, indicating the unavailability of certain definitions on the device, it would have been much faster to deal with.
Thanks!
